### PR TITLE
Stop ignoring many clippy lints.

### LINKED ISSUE #1287: Remove the allow clippy macro where it makes sense
It looks like we abused the usage of clippy allow macro when we first introduced the clippy integration test.

There are some clippy warnings that should be fixed instead.

The following are about style and making the code more rust like (these should be relatively easy to fix):
- needless_lifetimes
- eq_op
- identity_op
- needless_range_loop
- redundant_closure
- needless_lifetimes

These are more related to intrinsec design of Firecracker, and probably require more work:
- too_many_arguments
- cyclomatic_complexity
- large_enum_variant

To get all clippy allow macros, you can run:
```grep -rnI "allow(clippy"```

### PR DESCRIPTION
**Pull request by tamionv**
## Reason for This PR

#1287 , #1142
This does not fully fix #1287, but approximately halves the number of ignored clippy lints.

## Description of Changes

This PR does mainly stylistic refactoring in order to remove many occurrences ```#[allow(clippy::...)]```. The only meaningful change is a change in the definition of ```ParsedRequest```, boxing the largest enum variant. This will make our code more efficient, as the enum will no longer be forced to have its largest possible size at all times.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`  
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] Either this PR is linked to an issue, or, the reason for this PR is
      clearly provided. 
- [x] The description of changes is clear and encompassing.
- [x] Either no docs need to be updated as part of this PR, or, the required
      doc changes are included in this PR. Docs in scope are all `*.md` files
      located either in the repository root, or in the `docs/` directory.
- [x] Either no code has been touched, or, code-level documentation for touched
      code is included in this PR.
- [x] Either no API changes are included in this PR, or, the API changes are
      reflected in `firecracker/swagger.yaml`.
- [x] Either the changes in this PR have no user impact, or, the changes in
      this PR have user impact and have been added to the `CHANGELOG.md` file.

### CODE REVIEW

File: devices/src/legacy/i8042.rs [hunk 1]
```diff
@@ -63,8 +63,7 @@ const SB_I8042_CMD_DATA: u8 = 1 << 3; // i8042 expecting command parameter at po
 const SB_KBD_ENABLED: u8 = 1 << 4; // 1 = kbd enabled, 0 = kbd locked
 
 /// i8042 control register bits
-#[allow(clippy::identity_op)]
-const CB_KBD_INT: u8 = 1 << 0; // kbd interrupt enabled
+const CB_KBD_INT: u8 = 1; // kbd interrupt enabled
```
Comment (petreeftime) [inline devices/src/legacy/i8042.rs]: clippy is being stupid here and 1 << 0 is more expressive in this case.

Comment (tamionv) [reply devices/src/legacy/i8042.rs]: Okay sure, I will revert this -- but to optimize the number of CI calls I'll do a push when I get some more feedback

Comment (andreeaflorescu) [reply devices/src/legacy/i8042.rs]: Hmmm? What do you mean @petreeftime? Clippy is recommanding to **not** use 1 << 0 and just use 1 instead.

Comment (petreeftime) [reply devices/src/legacy/i8042.rs]: Agreed. But in this particular case this value means `bit 0`, which is better expressed as `1 << 0`. `1` by itself loses that meaning.

Comment (petreeftime) [reply devices/src/legacy/i8042.rs]: It's also inconsistent with the rest of the  `const`s in this file, which have similar meanings.

File: memory_model/src/guest_address.rs [hunk 2]
```diff
@@ -109,13 +109,12 @@ mod tests {
...
 
     #[test]
-    #[allow(clippy::eq_op)]
     fn cmp() {
-        let a = GuestAddress(0x300);
-        let b = GuestAddress(0x301);
-        assert!(a < b);
-        assert!(b > a);
-        assert!(!(a < a));
+        for i in 1..10 {
```
Comment (andreeaflorescu) [inline memory_model/src/guest_address.rs]: Why is this better than assert!(GuestAddress(0x300), GuestAddress(0x301))?

File: src/main.rs [hunk 3]
```diff
@@ -309,7 +308,9 @@ mod tests {
                     ("ERROR", "main.rs", "stack backtrace:"),
                     ("0:", "0x", "firecracker::main::"),
                 ],
-            ) || println!("Could not validate backtrace!\n {:?}", Backtrace::new()) != ()
+            ),
+            "Could not validate backtrace!\n {:?}",
```
Comment (andreeaflorescu) [inline src/main.rs]: I don't think this works. Check the commit message for why was this workaround introduced: https://github.com/firecracker-microvm/firecracker/commit/73f4eab

File: devices/src/virtio/block.rs [hunk 4]
```diff
@@ -219,13 +219,12 @@ impl Request {
...
     }
 
-    #[allow(clippy::ptr_arg)]
     fn execute<T: Seek + Read + Write>(
         &self,
         disk: &mut T,
         disk_nsectors: u64,
         mem: &GuestMemory,
-        disk_id: &Vec<u8>,
+        disk_id: &[u8],
```
Comment (andreeaflorescu) [inline devices/src/virtio/block.rs]: Already done in a merged commit.

File: devices/src/legacy/i8042.rs (hunk 1 above)
Comment (andreeaflorescu) [reply devices/src/legacy/i8042.rs]: If we do want to be consistent then we should define these flags in hex. That's also how they're defined in the kernel and in the key scan codes (below in the code).

How about his solution?

Comment (andreeaflorescu) [reply devices/src/legacy/i8042.rs]: https://elixir.bootlin.com/linux/latest/source/include/linux/i8042.h#L47

Comment (petreeftime) [reply devices/src/legacy/i8042.rs]: Looks good!

Comment (tamionv) [reply devices/src/legacy/i8042.rs]: Okay I will change them to hex

File: micro_http/src/request.rs [hunk 5]
```diff
@@ -327,10 +324,11 @@ mod tests {
         };
 
         let request_line = b"GET http://localhost/home HTTP/1.0";
-        match RequestLine::try_from(request_line) {
-            Ok(request) => assert_eq!(request, expected_request_line),
-            Err(_) => assert!(false),
-        };
+        if let Ok(request) = RequestLine::try_from(request_line) {
```
[the remark this answers is not in the document]
Comment (tamionv) [reply micro_http/src/request.rs]: Yes that makes more sense

File: memory_model/src/guest_address.rs (hunk 2 above)
Comment (tamionv) [reply memory_model/src/guest_address.rs]: Since this actually tests all possible orderings between the 2 operands of <.

File: src/main.rs (hunk 3 above)
Comment (tamionv) [reply src/main.rs]: Okay I've fixed it to a version that maintains behaviour.

**Pull request merged**